client/client.go

```diff
@@ -1115,91 +1113,120 @@ func (r *NotaryRepository) bootstrapClient(checkInitialized bool) (*TUFClient, e
+// Returns the public keys used in the rotation.
+// Create new key pair if neither private nor public key is provided.
+func (r *NotaryRepository) pubKeyListForRotation(role data.RoleName, serverManaged bool, privKeys []string, pubKeys data.KeyList) (pubKeyList data.KeyList, err error) {
 	var pubKey data.PublicKey
 
-	// If server manages the key being rotated, request a rotation and return the new key
-	if serverManaged {
+	// for each of the following 5 cases:
+	// validate and generate key pairs if necessary
+	switch {
+	case serverManaged:
+		// If server manages the key being rotated, request a rotation and return the new key
 		remote := r.getRemoteStore()
 		pubKey, err = rotateRemoteKey(role, remote)
-		pubKeyList = make(data.KeyList, 0, 1)
 		pubKeyList = append(pubKeyList, pubKey)
-		if err != nil {
-			return nil, fmt.Errorf("unable to rotate remote key: %s", err)
+		return pubKeyList, err
+	case len(pubKeys) == 0 && len(privKeys) == 0:
+		if pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey); err != nil {
+			return nil, fmt.Errorf("unable to generate key: %s", err)
 		}
-		return pubKeyList, nil
-	}
-
-	// If no new keys are passed in, we generate one
-	if len(newKeys) == 0 {
-		pubKeyList = make(data.KeyList, 0, 1)
-		pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey)
 		pubKeyList = append(pubKeyList, pubKey)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("unable to generate key: %s", err)
-	}
-
-	// If a list of keys to rotate to are provided, we add those
-	if len(newKeys) > 0 {
-		pubKeyList = make(data.KeyList, 0, len(newKeys))
-		for _, keyID := range newKeys {
-			pubKey = r.CryptoService.GetKey(keyID)
-			if pubKey == nil {
+	case len(pubKeys) == 0 && len(privKeys) != 0:
+		for _, keyID := range privKeys {
+			if pubKey = r.CryptoService.GetKey(keyID); pubKey == nil {
 				return nil, fmt.Errorf("unable to find key: %s", keyID)
 			}
 			pubKeyList = append(pubKeyList, pubKey)
 		}
+	case len(pubKeys) != 0 && len(privKeys) != 0:
+		if err = privateKeyIDsMatchPublicKeys(privKeys, pubKeys); err != nil {
+			return nil, err
+		}
+		pubKeyList = pubKeys
+	default:
+		pubKeyList = pubKeys
 	}
 
-	// Convert to certs (for root keys)
-	if pubKeyList, err = r.pubKeysToCerts(role, pubKeyList); err != nil {
-		return nil, err
+	return r.pubKeysToCerts(role, pubKeyList)
+}
+
+// privateKeyIDsMatchPublicKeys checks each KeyID to have a corresponding public key
+func privateKeyIDsMatchPublicKeys(privKIDs []string, pubKeys data.KeyList) error {
+	pubKeyIDs := make(map[string]bool)
+	for _, k := range pubKeys {
+		kid, err := utils.CanonicalKeyID(k)
+		if err != nil {
+			return fmt.Errorf("invalid public key: %v", err)
+		}
+		pubKeyIDs[kid] = true
 	}
 
-	return pubKeyList, nil
+	// check if each private key has corresponding public key
+	for _, id := range privKIDs {
+		if exist := pubKeyIDs[id]; !exist {
+			return fmt.Errorf("require a public key matching private key ID: %v", id)
```

Comment:
@riyazdf @endophage thanks for the feedback 😄 . The logic can definitely be made similar to the case `len(pubKeys) == 0 && len(privKeys) != 0`.

`privateKeyIDsMatchPublicKeys` confirms that all the private keys are provided with one matching public key. other unmatched public keys are assumed to be the alternate signing keys for other members of the organization (proposed by @endophage). 

I'm not sure if there are cases when we need more private keys than public keys. If there are then I can change it to accommodate that too.